astropy/cosmology/flrw.py

```diff
@@ -1485,6 +1454,99 @@ def arcsec_per_kpc_proper(self, z):
+
+    def __equiv__(self, other):
+        """FLRW equivalence. Use ``.is_equivalent()`` for actual check!
+
+        Parameters
+        ----------
+        other : `~astropy.cosmology.FLRW` subclass instance
+            The object in which to compare.
+
+        Returns
+        -------
+        `True` or `NotImplemented`
+            `True` if 'other' is of the same class and has matching parameters
+            and parameter values. `NotImplemented` otherwise.
+
+        """
+        # Note! this method is superseded by FlatFLRWMixin.__equiv__ in
+        # subclassses of FlatFLRWMixin.
+
+        if other.__class__ is not self.__class__:
+            return NotImplemented
+
+        # check all parameters in other match those in 'self' and 'other' has
+        # no extra parameters (latter part should never happen b/c same class)
+        sps, ops = self.info("parameters"), other.info("parameters")
+        params_eq = (all((np.all(ops[k] == v) for k, v in sps.items()))
+                     and not (sps.keys() ^ ops.keys()))
+
+        return params_eq or NotImplemented
+
+
+class FlatFLRWMixin(FlatCosmologyMixin):
+    """
+    Mixin class for flat FLRW cosmologies. Do NOT instantiate directly.
+    Must precede the base class in the multiple-inheritance so that this mixin's
+    ``__init__`` proceeds the base class'.
+    Note that all instances of ``FlatFLRWMixin`` are flat, but not all
+    flat cosmologies are instances of ``FlatFLRWMixin``. As example,
+    ``LambdaCDM`` **may** be flat (for the a specific set of parameter values),
+    but ``FlatLambdaCDM`` **will** be flat.
+    """
+
+    def __init__(self, *args, **kw):
+        super().__init__(*args, **kw)
+        # Do some twiddling after the fact to get flatness
+        self._Ode0 = 1.0 - self._Om0 - self._Ogamma0 - self._Onu0
+        self._Ok0 = 0.0
+
+    @property  # no longer a Parameter
+    def Ode0(self):
+        """Omega dark energy; dark energy density/critical density at z=0"""
+        return self._Ode0
+
+    def __equiv__(self, other):
+        """flat-FLRW equivalence. Use ``.is_equivalent()`` for actual check!
+
+        Parameters
+        ----------
+        other : `~astropy.cosmology.FLRW` subclass instance
+            The object in which to compare.
+
+        Returns
+        -------
+        `True` or `NotImplemented`
+            `True` if 'other' is of the same class / non-flat class (e.g.
+            ``FlatLambdaCDM`` and ``LambdaCDM``) has matching parameters
+            and parameter values. `NotImplemented` otherwise.
+
+        """
+        # check if case (1): same class & parameters
+        if isinstance(other, FlatFLRWMixin):
+            return super().__equiv__(other)
+
+        # check cases (3, 4), if other is the non-flat version of this class
+        # mro: [0: flat-class, 1: FlatFLRWMixin, 2: FlatCosmologyMixin, 3: non-flat]
+        elif other.__class__ is not self.__class__.mro()[3]:
```

Comment:
This seems super-fishy. How can you be absolutely sure with what subclasses/mixins a user might define?

Could one instead go by the keys being the same except for `Ode0`?

Or perhaps leave this to `other`? Would it not already automatically do the right thing if it just checked that `isinstance(flat_instance, my_general_class)`?

Reply:
> This seems super-fishy. How can you be absolutely sure with what subclasses/mixins a user might define?

Agreed. I really didn't like it. 

> Could one instead go by the keys being the same except for Ode0?

Unfortunately not. Having the same Parameters is no guarantee of having the same physics. It so happens for all the Cosmologies in this package, but e.g. a non-FLRW cosmo might have the same set of Parameters.

> Or perhaps leave this to ``other``? Would it not already automatically do the right thing if it just checked that ``isinstance(flat_instance, my_general_class)``?

I'm not sure I understand this. How does it know ``my_general_class``?

Reply:
Sorry, I should have been clearer: I thought `other` could check `self` is a subclass of `other`. But maybe that is illogical: no class has to know about its subclasses. On the other hand, if `self` finds here it is a subclass of `other`, then can it just use `other.__parameters__` as the attributes to check? (using `getattr` as suggested below)

Reply:
I implement a more stable class check.

```
comparable_classes = [c for c in self.__class__.mro()[1:]
                                        if (issubclass(c, FLRW) and c is not FLRW)]
```

Reply:
``FlatFLRWMixin.__equiv__`` is very similar to ``FLRW.__equiv__``, just more permissive in the classes and ``Ode0, Ok0`` parameters.  Now that ``is_equivalent`` checks both ``self.__equiv__`` and ``other.__equiv__``, this should work for ``self.is_equivalent(other)`` or ``other.is_equivalent(self)``.